pkg/networkserver/adr.go

```diff
@@ -138,22 +138,32 @@ func adaptDataRate(dev *ttnpb.EndDevice, fps *frequencyplans.Store, defaults ttn
 		return nil
 	}
 
-	maxSNR := ups[0].RxMetadata[0].SNR
-	for _, up := range ups {
+	var maxSNR float32
+	var i int
+	var up *ttnpb.UplinkMessage
+	for i, up = range ups {
+		if len(up.RxMetadata) > 0 {
+			maxSNR = up.RxMetadata[0].SNR
+			break
+		}
+	}
+	if i == len(ups)-1 && len(up.RxMetadata) == 0 {
+		return nil
+	}
+	for _, up = range ups[i:] {
 		for _, md := range up.RxMetadata {
 			if md.SNR > maxSNR {
 				maxSNR = md.SNR
 			}
 		}
 	}
+	// NOTE: up represents the last element in ups at this point.
```

Comment:
I had to read this three times to see what's going on and which variable means that and what its lifetime is. No immediate need for change, unless you have an idea to make this more verbose.